### ISSUE TITLE
SPAlternateUrl not working with Central Admin

### ISSUE THREAD

**Issue by ronx** [PowerShell/SharePointDsc#725]
**Details of the scenario you try and problem that is occurring:**
Use SPAlternateUrl to set HTTPS url for Central Admin's Intranet zone.

Looks like the Get-SPWebApplication just needs to -IncludeCentralAdministration

**The DSC configuration that is using the resource:**

**Version of the Operating System and PowerShell the DSC Target Node is running:**
WS2016, WMF 5.1

**Version of the DSC module you're using:**
2.0

**Comment by ronx** [comment]
Actually, even with -IncludeCentralAdministration, this doesn't work. I thought it was using the DisplayName property but it seems to be using the Name property instead. By default, these are the same in normal web applications. However, it appears that central administration has an empty Name property. I think what we need is something like:

`$webapp = Get-SPWebApplication -IncludeCentralAdministration | Where-Object { $_.DisplayName -eq $params.WebAppName }`

**Comment by ronx** [comment]
One more thing I just noticed. The return hash table from Get-TargetResource would also need to be changed to use $wa.DisplayName instead of $wa.Name for the WebAppName property.

Are there any cases we can think of where using DisplayName instead of Name would present issues?

**Comment by ykuijs** [comment]
You should use "Set-SPCentralAdministration -Port 443" to configure HTTPS on the the Central Admin site: www.harbar.net/archive/2013/02/13/Using-SSL-for-Central-Administration-with-SharePoint-2013.aspx

Will check the Name/DisplayName issue.

**Comment by ykuijs** [comment]
Or in our case SPFarm, but I just reviewed the resource and it looks like it is unable to change the port once it has been configured

**Comment by ronx** [comment]
You wouldn't happen to have a hard copy of that article, would you? That link has been broken for a while now, and I haven't been able to find a straightforward solution for this.

Using SPAlternateUrl seems to work well in combination with xWebsite to manage IIS bindings. What does Set-SPCentralAdministration do differently? Does it change the URL to use https for the default zone? I'm guessing hostname binding and cert selection still needs to be done through IIS (or corresponding DSC resources)? I'd just like to find the best way to implement this since I'm also adding it to the PLA. Let me know if you have any suggestions. Thanks!

**Comment by ykuijs** [comment]
The link works just fine for me. Have send a PDF copy of that page to you via mail.

Set-SPCentralAdministration configures the Central Admin both in the Config database, the server registry and IIS.

SPAlternateURL just updates the info in the Configuration database and you need xWebsite to update IIS.

**Comment by ykuijs** [comment]
Just checked the Name/DisplayName issue. The value of both properties are exactly the same, so we can use either one.

**Comment by ykuijs** [comment]
Was working on fixing this issue. Unfortunately I am running into an issue:
Because the Name property of the SPWebApplication object is empty, Get-SPWebApplication is unable to retrieve the Central Admin web application by using the name. This can be resolved by using Where-Object { $_.DisplayName -eq $params.WebAppName }.

Get-SPAlternateUrl however has exactly the same issue. Unfortunately, the Where-Object solution does not work for this cmdlet.

Anybody any clue how to resolve this issue??

**Comment by ronx** [comment]
How about Get-SPAlternateUrl -WebApplication $webapp

Does that give you what you need?

**Comment by ykuijs** [comment]
That is exactly what I was testing as well. First tests look promising!

**Comment by ronx** [comment]
I think the problem here is that in the existing SPAlternateUrl resource, it has no way to tell if you want to add a new internal URL or edit the existing one. Can anyone think of a case where someone would want to add a new internal URL for Central Administration rather than changing the existing one? If not, we should be able to knock this out without introducing any breaking changes.

**Comment by ykuijs** [comment]
Just submitted a PR which implements the possibility to change the Central Admin URL's. We need to extend this a little more, but wanted to release the PR for the v2.1 release.

Also:
When not using the Set-SPCentralAdministration cmdlet, the link to the CA site in the registry is not updated. This can be solved by using the Registry resource, but that needs to be done for all servers in the farm which is not desirable. Set-SPCentralAdministration will do this automatically on all servers.

**Comment by ronx** [comment]
Regarding the link to the CA site in the registry, I tested and Set-SPAlternateUrl also updates the registry. It seems to have the same effect as Set-SPCentralAdministration, except it doesn't touch IIS. It executes much faster than Set-SPCentralAdministration, so I'm guessing it's the IIS part that makes Set-SPCentralAdministration take so long.

In my fork that has the CA-specific handling, when SPAlternateUrl runs on only my Central Admin server, it executes Set-SPAlternateUrl, which sets the AAM for CA properly AND updates the CentralAdministrationUrl value on all servers. I didn't realize Set-SPAlternateUrl could do this, but I just tested it again manually and it does indeed work this way. At least, it does in SP2016. It would be worthwhile to test in SP2013 to make sure, but I still think my changes could be a workable solution for CA. Let me know what you think.
